clang/lib/Driver/ToolChains/CommonArgs.cpp

```diff
@@ -977,14 +977,61 @@ bool tools::addOpenMPRuntime(ArgStringList &CmdArgs, const ToolChain &TC,
   return true;
 }
 
-void tools::addFortranRuntimeLibs(const ToolChain &TC,
+void tools::addFortranRuntimeLibs(const ToolChain &TC, const ArgList &Args,
                                   llvm::opt::ArgStringList &CmdArgs) {
   // These are handled earlier on Windows by telling the frontend driver to add
   // the correct libraries to link against as dependents in the object file.
   if (!TC.getTriple().isKnownWindowsMSVCEnvironment()) {
+    // The --whole-archive option needs to be part of the link line to
+    // make sure that the main() function from Fortran_main.a is pulled
+    // in by the linker.  Determine if --whole-archive is active when
+    // flang will try to link Fortran_main.a.  If it is, don't add the
+    // --whole-archive flag to the link line.  If it's not, add a proper
+    // --whole-archive/--no-whole-archive bracket to the link line.
+    bool NeedWholeArchive = true;
+    auto * Arg = Args.getLastArg(options::OPT_Wl_COMMA);
```

Comment:
This can be `nullptr` if the -Wl option wasn't present, so we need to check for that.

Reply:
Right.  I'll fix this when I fixup the detecting algorithm.  The getLastArg thing was a red herring :-)

Reply:
Should be fixed, too.